File: src/test/java/com/dongtronic/diabot/util/PatternsTest.java
package com.dongtronic.diabot.util;

import org.junit.Assert;
import org.junit.Test;

import java.util.regex.Matcher;

public class PatternsTest {

  public PatternsTest() {
  }


  @Test
  public void unitStartNoSpacesMg() {
    String message = "100mg";

    Matcher matcher = Patterns.unitBgPattern.matcher(message);

    Assert.assertTrue(matcher.matches());

    Assert.assertEquals("100", matcher.group(4));
    Assert.assertEquals("mg", matcher.group(5));
  }

  @Test
  public void unitStartNoSpacesMmol() {
    String message = "5.5mmol";

    Matcher matcher = Patterns.unitBgPattern.matcher(message);

    Assert.assertTrue(matcher.matches());

    Assert.assertEquals("5.5", matcher.group(4));
    Assert.assertEquals("mmol", matcher.group(5));
  }

  @Test
  public void unitStartNoSpacesMgDecimal() {
    String message = "100.5mg";

    Matcher matcher = Patterns.unitBgPattern.matcher(message);

    Assert.assertTrue(matcher.matches());

    Assert.assertEquals("100.5", matcher.group(4));
    Assert.assertEquals("mg", matcher.group(5));
  }

  @Test
  public void unitStartNoSpacesMmolWhole() {
    String message = "5mmol";

    Matcher matcher = Patterns.unitBgPattern.matcher(message);

    Assert.assertTrue(matcher.matches());

    Assert.assertEquals("5", matcher.group(4));
    Assert.assertEquals("mmol", matcher.group(5));
  }

  @Test
  public void unitStartSpacesMg() {
    String message = "100 mg";

    Matcher matcher = Patterns.unitBgPattern.matcher(message);

    Assert.assertTrue(matcher.matches());

    Assert.assertEquals("100", matcher.group(4));
    Assert.assertEquals("mg", matcher.group(5));
  }

  @Test
  public void unitStartSpacesMmol() {
    String message = "5.5 mmol";

    Matcher matcher = Patterns.unitBgPattern.matcher(message);

    Assert.assertTrue(matcher.matches());

    Assert.assertEquals("5.5", matcher.group(4));
    Assert.assertEquals("mmol", matcher.group(5));
  }

  @Test
  public void unitStartSpacesMgDecimal() {
    String message = "100.5 mg";

    Matcher matcher = Patterns.unitBgPattern.matcher(message);

    Assert.assertTrue(matcher.matches());

    Assert.assertEquals("100.5", matcher.group(4));
    Assert.assertEquals("mg", matcher.group(5));
  }

  @Test
  public void unitStartSpacesMmolWhole() {
    String message = "5 mmol";

    Matcher matcher = Patterns.unitBgPattern.matcher(message);

    Assert.assertTrue(matcher.matches());

    Assert.assertEquals("5", matcher.group(4));
    Assert.assertEquals("mmol", matcher.group(5));
  }
  
  /////

  @Test
  public void textStartNoSpacesMg() {
    String message = "bla bla 100mg";

    Matcher matcher = Patterns.unitBgPattern.matcher(message);

    Assert.assertTrue(matcher.matches());

    Assert.assertEquals("100", matcher.group(4));
    Assert.assertEquals("mg", matcher.group(5));
  }

  @Test
  public void textStartNoSpacesMmol() {
    String message = "bla bla 5.5mmol";

    Matcher matcher = Patterns.unitBgPattern.matcher(message);

    Assert.assertTrue(matcher.matches());

    Assert.assertEquals("5.5", matcher.group(4));
    Assert.assertEquals("mmol", matcher.group(5));
  }

  @Test
  public void textStartNoSpacesMgDecimal() {
    String message = "bla bla 100.5mg";

    Matcher matcher = Patterns.unitBgPattern.matcher(message);

    Assert.assertTrue(matcher.matches());

    Assert.assertEquals("100.5", matcher.group(4));
    Assert.assertEquals("mg", matcher.group(5));
  }

  @Test
  public void textStartNoSpacesMmolWhole() {
    String message = "bla bla 5mmol";

    Matcher matcher = Patterns.unitBgPattern.matcher(message);

    Assert.assertTrue(matcher.matches());

    Assert.assertEquals("5", matcher.group(4));
    Assert.assertEquals("mmol", matcher.group(5));
  }

  @Test
  public void textStartSpacesMg() {
    String message = "bla bla 100 mg";

    Matcher matcher = Patterns.unitBgPattern.matcher(message);

    Assert.assertTrue(matcher.matches());

    Assert.assertEquals("100", matcher.group(4));
    Assert.assertEquals("mg", matcher.group(5));
  }

  @Test
  public void textStartSpacesMmol() {
    String message = "bla bla 5.5 mmol";

    Matcher matcher = Patterns.unitBgPattern.matcher(message);

    Assert.assertTrue(matcher.matches());

    Assert.assertEquals("5.5", matcher.group(4));
    Assert.assertEquals("mmol", matcher.group(5));
  }

  @Test
  public void textStartSpacesMgDecimal() {
    String message = "bla bla 100.5 mg";

    Matcher matcher = Patterns.unitBgPattern.matcher(message);

    Assert.assertTrue(matcher.matches());

    Assert.assertEquals("100.5", matcher.group(4));
    Assert.assertEquals("mg", matcher.group(5));
  }

  @Test
  public void textStartSpacesMmolWhole() {
    String message = "bla bla 5 mmol";

    Matcher matcher = Patterns.unitBgPattern.matcher(message);

    Assert.assertTrue(matcher.matches());

    Assert.assertEquals("5", matcher.group(4));
    Assert.assertEquals("mmol", matcher.group(5));
  }

  @Test
  public void negativeNoSpacesMmol() {
    String message = "-5.5mmol";

    Matcher matcher = Patterns.unitBgPattern.matcher(message);

    Assert.assertTrue(matcher.matches());

    Assert.assertEquals("5.5", matcher.group(4));
    Assert.assertEquals("mmol", matcher.group(5));
  }

  @Test
  public void negativeNoSpacesMmolWhole() {
    String message = "-5mmol";

    Matcher matcher = Patterns.unitBgPattern.matcher(message);

    Assert.assertTrue(matcher.matches());

    Assert.assertEquals("5", matcher.group(4));
    Assert.assertEquals("mmol", matcher.group(5));
  }

  @Test
  public void negativeNoSpacesMg() {
    String message = "-100mg";

    Matcher matcher = Patterns.unitBgPattern.matcher(message);

    Assert.assertTrue(matcher.matches());

    Assert.assertEquals("100", matcher.group(4));
    Assert.assertEquals("mg", matcher.group(5));
  }

  @Test
  public void negativeNoSpacesMgDecimal() {
    String message = "-100.5mg";

    Matcher matcher = Patterns.unitBgPattern.matcher(message);

    Assert.assertTrue(matcher.matches());

    Assert.assertEquals("100.5", matcher.group(4));
    Assert.assertEquals("mg", matcher.group(5));
  }

  @Test
  public void textStartNegativeMmol() {
    String message = "bla bla -5.5mmol";

    Matcher matcher = Patterns.unitBgPattern.matcher(message);

    Assert.assertTrue(matcher.matches());

    Assert.assertEquals("5.5", matcher.group(4));
    Assert.assertEquals("mmol", matcher.group(5));
  }

  @Test
  public void textStartNegativeMmolWhole() {
    String message = "bla bla -5mmol";

    Matcher matcher = Patterns.unitBgPattern.matcher(message);

    Assert.assertTrue(matcher.matches());

    Assert.assertEquals("5", matcher.group(4));
    Assert.assertEquals("mmol", matcher.group(5));
  }

  @Test
  public void textStartNegativeMg() {
    String message = "bla bla -100mg";

    Matcher matcher = Patterns.unitBgPattern.matcher(message);

    Assert.assertTrue(matcher.matches());

    Assert.assertEquals("100", matcher.group(4));
    Assert.assertEquals("mg", matcher.group(5));
  }

  @Test
  public void textStartNegativeMgDecimal() {
    String message = "bla bla -100.5mg";

    Matcher matcher = Patterns.unitBgPattern.matcher(message);

    Assert.assertTrue(matcher.matches());

    Assert.assertEquals("100.5", matcher.group(4));
    Assert.assertEquals("mg", matcher.group(5));
  }
}

File: src/test/java/com/dongtronic/diabot/util/PatternsTest.kt
package com.dongtronic.diabot.util

import org.junit.Assert
import org.junit.Test

import java.util.regex.Matcher

class PatternsTest {


    @Test
    fun unitStartNoSpacesMg() {
        val message = "100mg"

        val matcher = Patterns.unitBgPattern.matcher(message)

        Assert.assertTrue(matcher.matches())

        Assert.assertEquals("100", matcher.group(4))
        Assert.assertEquals("mg", matcher.group(5))
    }

    @Test
    fun unitStartNoSpacesMmol() {
        val message = "5.5mmol"

        val matcher = Patterns.unitBgPattern.matcher(message)

        Assert.assertTrue(matcher.matches())

        Assert.assertEquals("5.5", matcher.group(4))
        Assert.assertEquals("mmol", matcher.group(5))
    }

    @Test
    fun unitStartNoSpacesMgDecimal() {
        val message = "100.5mg"

        val matcher = Patterns.unitBgPattern.matcher(message)

        Assert.assertTrue(matcher.matches())

        Assert.assertEquals("100.5", matcher.group(4))
        Assert.assertEquals("mg", matcher.group(5))
    }

    @Test
    fun unitStartNoSpacesMmolWhole() {
        val message = "5mmol"

        val matcher = Patterns.unitBgPattern.matcher(message)

        Assert.assertTrue(matcher.matches())

        Assert.assertEquals("5", matcher.group(4))
        Assert.assertEquals("mmol", matcher.group(5))
    }

    @Test
    fun unitStartSpacesMg() {
        val message = "100 mg"

        val matcher = Patterns.unitBgPattern.matcher(message)

        Assert.assertTrue(matcher.matches())

        Assert.assertEquals("100", matcher.group(4))
        Assert.assertEquals("mg", matcher.group(5))
    }

    @Test
    fun unitStartSpacesMmol() {
        val message = "5.5 mmol"

        val matcher = Patterns.unitBgPattern.matcher(message)

        Assert.assertTrue(matcher.matches())

        Assert.assertEquals("5.5", matcher.group(4))
        Assert.assertEquals("mmol", matcher.group(5))
    }

    @Test
    fun unitStartSpacesMgDecimal() {
        val message = "100.5 mg"

        val matcher = Patterns.unitBgPattern.matcher(message)

        Assert.assertTrue(matcher.matches())

        Assert.assertEquals("100.5", matcher.group(4))
        Assert.assertEquals("mg", matcher.group(5))
    }

    @Test
    fun unitStartSpacesMmolWhole() {
        val message = "5 mmol"

        val matcher = Patterns.unitBgPattern.matcher(message)

        Assert.assertTrue(matcher.matches())

        Assert.assertEquals("5", matcher.group(4))
        Assert.assertEquals("mmol", matcher.group(5))
    }

    /////

    @Test
    fun textStartNoSpacesMg() {
        val message = "bla bla 100mg"

        val matcher = Patterns.unitBgPattern.matcher(message)

        Assert.assertTrue(matcher.matches())

        Assert.assertEquals("100", matcher.group(4))
        Assert.assertEquals("mg", matcher.group(5))
    }

    @Test
    fun textStartNoSpacesMmol() {
        val message = "bla bla 5.5mmol"

        val matcher = Patterns.unitBgPattern.matcher(message)

        Assert.assertTrue(matcher.matches())

        Assert.assertEquals("5.5", matcher.group(4))
        Assert.assertEquals("mmol", matcher.group(5))
    }

    @Test
    fun textStartNoSpacesMgDecimal() {
        val message = "bla bla 100.5mg"

        val matcher = Patterns.unitBgPattern.matcher(message)

        Assert.assertTrue(matcher.matches())

        Assert.assertEquals("100.5", matcher.group(4))
        Assert.assertEquals("mg", matcher.group(5))
    }

    @Test
    fun textStartNoSpacesMmolWhole() {
        val message = "bla bla 5mmol"

        val matcher = Patterns.unitBgPattern.matcher(message)

        Assert.assertTrue(matcher.matches())

        Assert.assertEquals("5", matcher.group(4))
        Assert.assertEquals("mmol", matcher.group(5))
    }

    @Test
    fun textStartSpacesMg() {
        val message = "bla bla 100 mg"

        val matcher = Patterns.unitBgPattern.matcher(message)

        Assert.assertTrue(matcher.matches())

        Assert.assertEquals("100", matcher.group(4))
        Assert.assertEquals("mg", matcher.group(5))
    }

    @Test
    fun textStartSpacesMmol() {
        val message = "bla bla 5.5 mmol"

        val matcher = Patterns.unitBgPattern.matcher(message)

        Assert.assertTrue(matcher.matches())

        Assert.assertEquals("5.5", matcher.group(4))
        Assert.assertEquals("mmol", matcher.group(5))
    }

    @Test
    fun textStartSpacesMgDecimal() {
        val message = "bla bla 100.5 mg"

        val matcher = Patterns.unitBgPattern.matcher(message)

        Assert.assertTrue(matcher.matches())

        Assert.assertEquals("100.5", matcher.group(4))
        Assert.assertEquals("mg", matcher.group(5))
    }

    @Test
    fun textStartSpacesMmolWhole() {
        val message = "bla bla 5 mmol"

        val matcher = Patterns.unitBgPattern.matcher(message)

        Assert.assertTrue(matcher.matches())

        Assert.assertEquals("5", matcher.group(4))
        Assert.assertEquals("mmol", matcher.group(5))
    }

    @Test
    fun negativeNoSpacesMmol() {
        val message = "-5.5mmol"

        val matcher = Patterns.unitBgPattern.matcher(message)

        Assert.assertTrue(matcher.matches())

        Assert.assertEquals("5.5", matcher.group(4))
        Assert.assertEquals("mmol", matcher.group(5))
    }

    @Test
    fun negativeNoSpacesMmolWhole() {
        val message = "-5mmol"

        val matcher = Patterns.unitBgPattern.matcher(message)

        Assert.assertTrue(matcher.matches())

        Assert.assertEquals("5", matcher.group(4))
        Assert.assertEquals("mmol", matcher.group(5))
    }

    @Test
    fun negativeNoSpacesMg() {
        val message = "-100mg"

        val matcher = Patterns.unitBgPattern.matcher(message)

        Assert.assertTrue(matcher.matches())

        Assert.assertEquals("100", matcher.group(4))
        Assert.assertEquals("mg", matcher.group(5))
    }

    @Test
    fun negativeNoSpacesMgDecimal() {
        val message = "-100.5mg"

        val matcher = Patterns.unitBgPattern.matcher(message)

        Assert.assertTrue(matcher.matches())

        Assert.assertEquals("100.5", matcher.group(4))
        Assert.assertEquals("mg", matcher.group(5))
    }

    @Test
    fun textStartNegativeMmol() {
        val message = "bla bla -5.5mmol"

        val matcher = Patterns.unitBgPattern.matcher(message)

        Assert.assertTrue(matcher.matches())

        Assert.assertEquals("5.5", matcher.group(4))
        Assert.assertEquals("mmol", matcher.group(5))
    }

    @Test
    fun textStartNegativeMmolWhole() {
        val message = "bla bla -5mmol"

        val matcher = Patterns.unitBgPattern.matcher(message)

        Assert.assertTrue(matcher.matches())

        Assert.assertEquals("5", matcher.group(4))
        Assert.assertEquals("mmol", matcher.group(5))
    }

    @Test
    fun textStartNegativeMg() {
        val message = "bla bla -100mg"

        val matcher = Patterns.unitBgPattern.matcher(message)

        Assert.assertTrue(matcher.matches())

        Assert.assertEquals("100", matcher.group(4))
        Assert.assertEquals("mg", matcher.group(5))
    }

    @Test
    fun textStartNegativeMgDecimal() {
        val message = "bla bla -100.5mg"

        val matcher = Patterns.unitBgPattern.matcher(message)

        Assert.assertTrue(matcher.matches())

        Assert.assertEquals("100.5", matcher.group(4))
        Assert.assertEquals("mg", matcher.group(5))
    }
}
